src/bin/basic_subclass.rs

```diff
@@ -0,0 +1,176 @@
+            .expect("Failed to create SimpleAppWindow")
+            .downcast::<SimpleAppWindow>()
+            .expect("Created simpleappwindow is of wrong type");
+
+        let headerbar = gtk::HeaderBar::new();
+        headerbar.set_title(Some("Hello World!"));
+        headerbar.set_show_close_button(true);
+
+        let label = gtk::Label::new(Some("Hello World!"));
+
+        self_.add(&label);
+        self_.set_titlebar(Some(&headerbar));
+        self_.set_default_size(640, 480);
+
+        let priv_ = SimpleAppWindowPrivate::from_instance(&self_);
+        *priv_.state.borrow_mut() = Some(WindowState { headerbar, label });
+
+        self_
+    }
+}
+
+#[derive(Debug)]
+struct State {
+    window: SimpleAppWindow,
+}
+
+pub struct SimpleApplicationPrivate {
+    state: RefCell<Option<State>>,
+}
+
+impl ObjectSubclass for SimpleApplicationPrivate {
+    const NAME: &'static str = "SimpleApplicationPrivate";
+    type ParentType = gtk::Application;
+    type Instance = subclass::simple::InstanceStruct<Self>;
+    type Class = subclass::simple::ClassStruct<Self>;
+
+    glib_object_subclass!();
+
+    fn new() -> Self {
+        Self {
+            state: RefCell::new(None),
+        }
+    }
+}
+
+impl ObjectImpl for SimpleApplicationPrivate {
+    glib_object_impl!();
+}
+
+impl ApplicationImpl for SimpleApplicationPrivate {
+    fn activate(&self, app: &gio::Application) {
+        let app = app.downcast_ref::<gtk::Application>().unwrap();
+        let priv_ = SimpleApplicationPrivate::from_instance(app);
+        let mut state = priv_.state.borrow_mut();
+        if let Some(ref s) = *state {
+            s.window.present();
+            println!("presented!");
+        } else {
+            println!("stuff created!");
+            let window = SimpleAppWindow::new(&app);
+            *state = Some(State {
+                window: window.clone(),
+            });
+            window.show_all();
+            window.present();
+        }
+    }
+}
+
+impl GtkApplicationImpl for SimpleApplicationPrivate {}
+
+glib_wrapper! {
+    pub struct SimpleApplication(
+        Object<subclass::simple::InstanceStruct<SimpleApplicationPrivate>,
+        subclass::simple::ClassStruct<SimpleApplicationPrivate>,
+        SimpleApplicationClass>)
+        @extends gio::Application, gtk::Application;
+
+    match fn {
+        get_type => || SimpleApplicationPrivate::get_type().to_glib(),
+    }
+}
+
+impl SimpleApplication {
+    pub fn new() -> Self {
+        glib::Object::new(
+            Self::static_type(),
+            &[
+                ("application-id", &"org.gtk-rs.SimpleApplication"),
+                ("flags", &ApplicationFlags::empty()),
+            ],
+        )
+        .expect("Failed to create SimpleApp")
+        .downcast()
+        .expect("Created simpleapp is of wrong type")
+    }
+}
+
+fn main() {
+    // FIXME: shoudln't panic without this
```

Comment:
How does it panic exactly? (Also typo: shouldn't not shoudln't)

Reply:
```

Reply:
Here's what happens when commenting out the line

```
thread 'main' panicked at 'GTK has not been initialized. Call `gtk::init` first.', /home/alatiera/.cargo/git/checkouts/gtk-b89af3a825b1a0bb/91968e9/src/auto/header_bar.rs:38:9
stack backtrace:
   0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
             at src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:39
   1: std::sys_common::backtrace::_print
             at src/libstd/sys_common/backtrace.rs:71
   2: std::panicking::default_hook::{{closure}}
             at src/libstd/sys_common/backtrace.rs:59
             at src/libstd/panicking.rs:197
   3: std::panicking::default_hook
             at src/libstd/panicking.rs:211
   4: std::panicking::rust_panic_with_hook
             at src/libstd/panicking.rs:474
   5: std::panicking::begin_panic
             at /rustc/a53f9df32fbb0b5f4382caaad8f1a46f36ea887c/src/libstd/panicking.rs:408
   6: gtk::auto::header_bar::HeaderBar::new
             at /home/alatiera/.cargo/git/checkouts/gtk-b89af3a825b1a0bb/91968e9/src/auto/header_bar.rs:38
   7: basic_subclass::SimpleAppWindow::new
             at src/bin/basic_subclass.rs:73
   8: <basic_subclass::SimpleApplicationPrivate as gio::subclass::application::ApplicationImpl>::activate
             at src/bin/basic_subclass.rs:128
   9: gio::subclass::application::application_activate
             at /home/alatiera/.cargo/git/checkouts/gio-a4b825b310a073b9/9a06d05/src/subclass/application.rs:313
  10: <unknown>
  11: g_signal_emit_valist
  12: g_signal_emit
  13: <unknown>
  14: <T as gio::subclass::application::ApplicationImplExt>::parent_local_command_line
             at /home/alatiera/.cargo/git/checkouts/gio-a4b825b310a073b9/9a06d05/src/subclass/application.rs:211
  15: gio::subclass::application::ApplicationImpl::local_command_line
             at /home/alatiera/.cargo/git/checkouts/gio-a4b825b310a073b9/9a06d05/src/subclass/application.rs:104
  16: gio::subclass::application::application_local_command_line
             at /home/alatiera/.cargo/git/checkouts/gio-a4b825b310a073b9/9a06d05/src/subclass/application.rs:366
  17: g_application_run
  18: <O as gio::application::ApplicationExtManual>::run
             at /home/alatiera/.cargo/git/checkouts/gio-a4b825b310a073b9/9a06d05/src/application.rs:23
  19: basic_subclass::main
             at src/bin/basic_subclass.rs:175
  20: std::rt::lang_start::{{closure}}
             at /rustc/a53f9df32fbb0b5f4382caaad8f1a46f36ea887c/src/libstd/rt.rs:64
  21: std::panicking::try::do_call
             at src/libstd/rt.rs:49
             at src/libstd/panicking.rs:293
  22: __rust_maybe_catch_panic
             at src/libpanic_unwind/lib.rs:85
  23: std::rt::lang_start_internal
             at src/libstd/panicking.rs:272
             at src/libstd/panic.rs:394
             at src/libstd/rt.rs:48
  24: std::rt::lang_start
             at /rustc/a53f9df32fbb0b5f4382caaad8f1a46f36ea887c/src/libstd/rt.rs:64
  25: main
  26: __libc_start_main
  27: _start
```

Reply:
Hrm, because `gtk::init()` is called by running the application but we don't know that at the Rust layer. There's an open issue about that so don't worry, just keep it and remove the FIXME comment.